File: src/Pawchums.Domain/Entities/RequestRescues/RequestRescue.cs
using AnimalRescueSystem.Entities.RequestRescues;
using System;
using System.Collections.Generic;
using Volo.Abp.Domain.Entities.Auditing;
using Volo.Abp.MultiTenancy;

namespace Pawchums.Entities.RequestRescues;

public class RequestRescue : FullAuditedAggregateRoot<Guid>, IMultiTenant
{
    public Guid? TenantId { get; set; }

    // Request Information (Posted by anyone - can be anonymous)
    public virtual string Title { get; set; }
    public virtual string Location { get; set; }
    public virtual string Description { get; set; }
    public virtual string? Picture { get; set; }
    public virtual string ContactNo { get; set; }
    public virtual string? ContactName { get; set; }
    public virtual DateTime RequestDate { get; set; }

    // Status Management
    public virtual string Status { get; set; } // NotInitiated, Initiated, InProgress, Completed, Cancelled

    // Severity
    public virtual string Severity { get; set; } // Low, Medium, High

    // Active Status
    public virtual bool IsActive { get; set; }
    public virtual double? Latitude { get; set; }
    public virtual double? Longitude { get; set; }
    public virtual string? MapUrl { get; set; }
    // Navigation Properties
    public virtual ICollection<RescueInitiation> RescueInitiations { get; set; }
    public virtual RescueCompletion? RescueCompletion { get; set; }

    public RequestRescue()
    {
        RescueInitiations = new List<RescueInitiation>();
    }
}

File: src/types/requestRescue.ts
// Request Rescue DTOs and Types

export interface CreateUpdateRequestRescueDto {
  title: string;
  location: string;
  description: string;
  picture?: string;
  contactNo: string;
  contactName?: string;
  isActive?: boolean;
}

export interface RequestRescueDto {
  id: string;
  title: string;
  location: string;
  description: string;
  picture?: string;
  contactNo: string;
  contactName?: string;
  requestDate: string;
  status: RequestRescueStatus;
  isActive: boolean;
  creationTime: string;
  initiationsCount: number;
  hasCompletion: boolean;
}

export interface RequestRescueResponseDto {
  id: string;
  title: string;
  status: string;
}

export interface RequestRescueFilter {
  searchKeyword?: string;
  status?: RequestRescueStatus;
  isActive?: boolean;
}

export type RequestRescueStatus =
  | 'NotInitiated'
  | 'Initiated'
  | 'InProgress'
  | 'Completed'
  | 'Cancelled';

export const REQUEST_RESCUE_STATUS_OPTIONS = [
  { label: 'Not Initiated', value: 'NotInitiated' },
  { label: 'Initiated', value: 'Initiated' },
  { label: 'In Progress', value: 'InProgress' },
  { label: 'Completed', value: 'Completed' },
  { label: 'Cancelled', value: 'Cancelled' },
] as const;
